go-controller/pkg/node/base_node_network_controller.go

```diff
@@ -15,38 +16,73 @@ import (
 	"k8s.io/klog/v2"
 )
 
-// updateIsOvnUpEnabled checks if ovnkube-master has been upgraded from OvnPortBindingTopoVersion
-// and start to support port binding up field. If so update atomicOvnUpEnabled accordingly.
-func (bnnc *BaseNodeNetworkController) updateIsOvnUpEnabled(ctx context.Context) error {
-	isOvnUpEnabled := atomic.LoadInt32(&bnnc.atomicOvnUpEnabled) > 0
-	if config.OvnKubeNode.Mode != types.NodeModeDPU || isOvnUpEnabled || config.OvnKubeNode.DisableOVNIfaceIdVer {
-		return nil
-	}
+// initUpgradeController creates the upgradeController and return the initial topologyVersion
+func (bnnc *BaseNodeNetworkController) initUpgradeController(ctx context.Context) (*upgrade.UpgradeController, int, error) {
```

Comment:
Why are `initUpgradeController`, `updateIsOvnUpEnabledDPUMode` and `updateIsOvnUpEnabledCommon` in `BaseNodeNetworkController` if they are only to be used from `DefaultNodeNetworkController`?

Reply:
so network controller manager to start a go routine to check  the upgrader's topolgogy version to be latest one and then update a global OVN IsOvnUpEnabled variable?

each DPU mode network controller would check that global variable when doing watch watchPodsDPU()?

as full mode network controller needs to update cniserver, it would still do its own check? or it starts a go routine to check  the glocal OVN IsOvnUpEnabled  is updated then update cniserver?

Is my understanding correct?

Reply:
Yes, although it doesn't need to be a global variable. That can be shared around without the need of a global variable.

Why would you like to do IsOvnUpEnabled per network controller? It's a global concept that cannot change at runtime and it wouldn't have different values per controller.

It looks to me that several things could be done better in an upper layer like node network controller manager rather than in network controllers. For example, the cni server could be handled there as well. And perhaps it should include part of the upgrade coordination. For example, network controllers could be initialized with the initial topology version and delay running until after the topology version has changed. Then you could run them passing some global topology dependant parameters like IsOvnUpEnabled.

We can improve this later but it certainly seems a waste to be doing the same thing in all controllers. I think we also had a related discussion for handling topology upgrades in master side and the idea I explained above could also apply there.

Reply:
I think it is much more complicated to make network_manager do it, it needs a general way to notify (some kind of registered function?) its managed networkControllers. It is either notify once to the whole managed controllers module, or notify one by one to all managed controllers, the latter needs some kind of lock in network manager which does not exist today. The former would be more complicated, each network controller needs to consume that same single notification. 

Or maybe you are thinking of a better way than mine, we can certainly disucss this in the next meeting.

Reply:
This other suggestion might be more workable:

```
For example, network controllers could be initialized with the initial topology version and delay running them until after the topology version has changed. Then you could run them passing some global topology dependent parameters like IsOvnUpEnabled.
```

But we can do that later.